pkg/controller/sparkapplication/controller.go

```diff
@@ -289,397 +300,472 @@ func (c *Controller) processNextItem() bool {
-	original *v1alpha1.SparkApplication,
-	toUpdate *v1alpha1.SparkApplication) (*v1alpha1.SparkApplication, error) {
-	if reflect.DeepEqual(original.Status, toUpdate.Status) {
-		return nil, nil
-	}
-
-	return c.crdClient.SparkoperatorV1alpha1().SparkApplications(toUpdate.Namespace).Update(toUpdate)
-}
-
-func (c *Controller) getSparkApplication(namespace string, name string) (*v1alpha1.SparkApplication, error) {
-	return c.lister.SparkApplications(namespace).Get(name)
-}
 
-// handleRestart handles application restart if the application has terminated and is subject to restart according to
-// the restart policy.
-func (c *Controller) handleRestart(app *v1alpha1.SparkApplication) {
-	glog.Infof("SparkApplication %s failed or terminated, restarting it with RestartPolicy %s",
-		app.Name, app.Spec.RestartPolicy)
+		if app.Spec.Monitoring != nil && app.Spec.Monitoring.Prometheus != nil {
+			// configPrometheusMonitoring may update app.Spec.
+			configPrometheusMonitoring(app, c.kubeClient)
+		}
 
-	// Delete the old driver pod and UI service if necessary. Note that in case an error occurred here, we simply
-	// log the error and continue enqueueing the application for resubmission because the driver has already
-	// terminated so failure to cleanup the old driver pod and UI service is not a blocker. Also note that because
-	// deleting a already terminated driver pod won't trigger a driver state update by the sparkPodMonitor so won't
-	// cause repetitive restart handling.
-	if err := c.deleteDriverAndUIService(app, false); err != nil {
-		glog.Errorf("failed to delete the old driver pod and UI service for SparkApplication %s: %v",
-			app.Name, err)
+		// Create Spark UI Service.
+		service, err := createSparkUIService(app, c.kubeClient)
+		if err != nil {
+			glog.Errorf("Failed to create UI service for SparkApplication %s: %v", app.Name, err)
+		} else {
+			app.Status.DriverInfo.WebUIServiceName = service.serviceName
+			app.Status.DriverInfo.WebUIPort = service.nodePort
+			// Create UI Ingress if ingress-format is set.
+			if c.ingressUrlFormat != "" {
+				ingress, err := createSparkUIIngress(app, *service, c.ingressUrlFormat, c.kubeClient)
+				if err != nil {
+					glog.Errorf("Failed to create UI Ingress for SparkApplication %s: %v", app.Name, err)
+				} else {
+					app.Status.DriverInfo.WebUIIngressAddress = ingress.ingressUrl
+					app.Status.DriverInfo.WebUIIngressName = ingress.ingressName
+				}
+			}
+		}
 	}
-
-	//Enqueue the object for re-submission.
-	c.enqueue(app)
-
-	c.recorder.Eventf(
-		app,
-		apiv1.EventTypeNormal,
-		"SparkApplicationRestart",
-		"SparkApplication %s was enqueued for restart",
-		app.Name)
+	return app
 }
 
-func (c *Controller) handleResubmission(app *v1alpha1.SparkApplication, submissionRetries int32) {
-	glog.Infof("Retrying submission of SparkApplication %s", app.Name)
+func (c *Controller) updateAppAndExportMetrics(oldApp, newApp *v1alpha1.SparkApplication) error {
```

Comment:
There's also the `updateSparkApplicationStatusWithRetries`, why you also do update here?

Reply:
The problem is this function doesn't really take care of update conflicts, whereas `updateSparkApplicationStatusWithRetries` does. I think we should combine the two functions in some way to avoid confusion.

Reply:
I still don’t think we should do retries here since our state-machine is idempotent and eventually consistent i.e. if we process a CRD based on our state-machine transitions and fail to write to etc.d because of update conflicts, the next time we process the CRD (which has an upper bound and at-worst can be equal our informer cache re-sync period), the state-machine will process the CRD correctly and drive it to its desired state.

Secondly, adding retries here can lead to a race-condition with our CRD update processing. For e.g. If a user updates the CRD while a worker is processing the SparkApplication, the onUpdate callback will try to update the CRD status to Invalidating but the worker once its done processing can potentially overwrite this Invalidating state if we add retries here leading to an onUpdate miss.  In the current implementation, the worker overwrite will fail and the Invalidating state will be processed on the next time we process the CRD.  

Let me know what you think on this.

Reply:
Also made some minor changes to how we clean up resources on failures/re-run etc. to make sure this is true.